### PR TITLE
Fixing deadlocks and simplifying code by using concurrent collections

### LINKED ISSUE #80: Unknown server error resulted in premature abort of console thread
**Describe the bug**
I left the server on, it must have crashed during the night, I went to check the error logs, and I had this

**Version Encountered**

**To Reproduce**
Just leave the server open.

**Logs**
```
2020-04-03 22:18:44.815 [Error] 
  Message: Thread was being aborted.
    Stack:    at System.IO.__ConsoleStream.WaitForAvailableConsoleInput(SafeFileHandle file, Boolean isPipe)
   at System.IO.__ConsoleStream.ReadFileNative(SafeFileHandle hFile, Byte[] bytes, Int32 offset, Int32 count, Boolean useFileAPIs, Boolean isPipe, Int32& bytesRead)
   at System.IO.__ConsoleStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.StreamReader.ReadBuffer()
   at System.IO.StreamReader.ReadLine()
   at System.IO.TextReader.SyncTextReader.ReadLine()
   at System.Console.ReadLine()
   at Intersect.Console.ReadLine() in G:\Celtos Online\Intersect-Engine\Intersect (Core)\Console.cs:line 362
   at Intersect.Server.Core.ServerConsole.ThreadStart()
```

**Operating System:**
Windows Server 2016

[error.txt](https://github.com/AscensionGameDev/Intersect-Engine/files/4432198/error.txt)

### PR DESCRIPTION
**Pull request by jcsnider**
Resolves #80

### CODE REVIEW

File: Intersect.Server/Maps/MapInstance.cs [hunk 1]
```diff
@@ -329,60 +331,33 @@ public void DespawnItems()
 
         public void DespawnNpcsOf(NpcBase npcBase)
         {
-            var npcs = new List<Npc>();
-            lock (GetMapLock())
+            foreach (var entity in mEntities.Values)
             {
-                foreach (var entity in mEntities)
+                if (entity.GetType() == typeof(Npc) && ((Npc)entity).Base == npcBase)
```
Comment (lodicolo) [inline Intersect.Server/Maps/MapInstance.cs:336]: This should be `if (entity is Npc npc && npc.Base == npcBase)`.

File: Intersect.Server/Maps/MapInstance.cs [hunk 2]
```diff
@@ -329,60 +331,33 @@ public void DespawnItems()
...
-                foreach (var entity in mEntities)
-                {
-                    if (entity.GetType() == typeof(Resource) && ((Resource) entity).Base == resourceBase)
-                    {
-                        resources.Add((Resource) entity);
-                    }
-                }
-
-                foreach (var en in resources)
+                if (entity.GetType() == typeof(Resource) && ((Resource)entity).Base == resourceBase)
```
Comment (lodicolo) [inline Intersect.Server/Maps/MapInstance.cs:347]: This should be `if (entity is Resource resource && resource.Base == resourceBase)`.

File: Intersect.Server/Maps/MapInstance.cs [hunk 3]
```diff
@@ -329,60 +331,33 @@ public void DespawnItems()
...
-                foreach (var entity in mEntities)
-                {
-                    if (entity.GetType() == typeof(Projectile) && ((Projectile) entity).Base == projectileBase)
-                    {
-                        projectiles.Add((Projectile) entity);
-                    }
-                }
-
-                foreach (var en in projectiles)
+                if (entity.GetType() == typeof(Projectile) && ((Projectile)entity).Base == projectileBase)
```
Comment (lodicolo) [inline Intersect.Server/Maps/MapInstance.cs:358]: This should be `if (entity is Projectile projectile && projectile.Base == projectileBase)`.

File: Intersect.Server/Maps/MapInstance.cs [hunk 4]
```diff
@@ -760,21 +728,11 @@ public void RemoveTrap(MapTrapInstance trap)
 
         public void ClearEntityTargetsOf(Entity en)
         {
-            if (Monitor.TryEnter(GetMapLock(), new TimeSpan(0, 0, 0, 0, 1)))
+            foreach (var entity in mEntities.Values)
             {
-                try
+                if (entity.GetType() == typeof(Npc) && ((Npc)entity).Target == en)
```
Comment (lodicolo) [inline Intersect.Server/Maps/MapInstance.cs:733]: This should be `if (entity is Npc npc && npc.Target == en)`.

File: Intersect.Server/Maps/MapInstance.cs [hunk 5]
```diff
@@ -828,22 +786,22 @@ public void Update(long timeMs)
...
-                    for (var i = 0; i < mEntities.Count; i++)
+                    foreach (var en in mEntities.Values)
                     {
                         //Let's see if and how long this map has been inactive, if longer than 30 seconds, regenerate everything on the map
                         //TODO, take this 30 second value and throw it into the server config after I switch everything to json
                         if (timeMs > LastUpdateTime + 30000)
                         {
                             //Regen Everything & Forget Targets
-                            if (mEntities[i].GetType() == typeof(Resource) || mEntities[i].GetType() == typeof(Npc))
+                            if (en.GetType() == typeof(Resource) || en.GetType() == typeof(Npc))
```
Comment (lodicolo) [inline Intersect.Server/Maps/MapInstance.cs:796]: This should be `if (en is Resource || en is Npc)`.

**Comment by jcsnider** [comment]
@lodicolo every single .GetType() has been removed from that class for ya :)

**Pull request merged**